README.md
# mini-mock-api
Lightweight node.js API mock server, route calls to static JSON files

## What does it do?

Mini-Mock-API was written to mock a simple API in a few minutes. API calls are routed to static JSON files.
As an example if your mock directory looks like this:
```
+ mock-files
  - cars.json
  - trains.json
  + oldschool
    - carriages.json
```

with cars.json
```
[
  {id: 1, name: 'Porsche 911'},
  {id: 2, name: 'Tesla S'}
]
```

`GET localhost:8080/api/v1/cars`

returns all cars inside cars.json

`GET localhost:8080/api/v1/cars/1`

returns one car with the id == 1

`GET localhost:8080/api/v1/oldschool/carriages`

returns all items from oldschool/carriages.json


## Installation

`npm install mini-mock-api`

## Usage

```
var API = require('mini-mock-api');
var myApi = new API({
  basePath: '/api/v1',
  mockPath: 'mock-files',
  idAttribute: '_id'
});
myApi.start();
```

To run the server go to the script directory (path options are relative to your execution path):
```
cd examples
node example.js
```

Open your browser and go to `localhost:8080/api/v1/todos`

## Options
```
var options = {
  mockPath: 'mocks', //directory to mock files relative to path of script
  basePath: '/api/v1',  //base path for api calls -> localhost:8080/api/v1/...
  port: 8080,
  idAttribute: 'uuid', //the id property to search for when requesting api/v1/cars/123
  cors: false, //CORS headers for cross origin requests
  sortParameter: 'sortOrder' //the GET parameter for sort orders e.g. api/v1/cars?sortOrder=+name
  delay: false //true adds a 500ms delay to all request using connect-slow (http://npmjs.com/package/connect-slow), you can also set a options object to use connect-slow's advanced features
}
var myApi = new API(options);
```

## Custom Routes

If static JSON files are not sufficient you can also add custom routes by defining your own request handler functions. Custom routes are supported for GET, POST, PUT and DELETE methods:
```
myApi.get('custom/status', function(request, response){
  response.json({status: 'okay'});
});
```

`-> localhost:8080/api/v1/custom/status`

Please note that custom routes are relativ to your 'basePath' property. If you need a route outside of your API scope use the '[get|post|put|delete]FromRoot' method:
```
myApi.getFromRoot('/ping', function(request, response){
  response.json({hello: 'there'});
});
```

`-> localhost:8080/ping`

## Decorate

All API responses can be decorated by defining a `decorate` function. The functions also gets passed the request object as the second parameter. In this example all returned collections are wrapped into `results` and a total count is added.

```
myApi.decorate = function(data, request){
  if(data.length){
    return {
      results: data,
      total: data.length
    };
  }
  return data;
};
```

## Develop

* Install dependencies with ```npm install```
* Run tests with ```jasmine```



## Can i use this in production?

Hell no!


CHANGELOG.md
<a name="0.3.0"></a>

# 0.3.0 (2015-09-02)

## Features

- Add basic tests


<a name="0.2.3"></a>

# 0.2.3 (2015-08-20)

## Features

- Add delay option using connect-slow ([9453110](https://github.com/pwambach/mini-mock-api/commit/9453110f1cde1064a220dcf2ac0d1b2828facdff))



<a name="0.2.2"></a>

# 0.2.2 (2015-07-21)

## Bug Fixes

- log exception from decorate function ([8bcadce](https://github.com/pwambach/mini-mock-api/commit/8bcadce7f2ed63c88a27540e8ac34db4cdada9a4))


# 0.2.1 (2015-06-29)

## Features

- add CORS headers ([#1](https://github.com/pwambach/mini-mock-api/issues/1)) thx @TheBox193
- request object as second parameter in decorate function ([a9c8eeb](https://github.com/pwambach/mini-mock-api/commit/a9c8eebdd80047158fab71c3cfbde092aac6ba5e))


<a name="0.2.0"></a>
# 0.2.0 (2015-06-29)

## Features

- custom routes ([764d1c3](https://github.com/pwambach/mini-mock-api/commit/764d1c3975136e880023c85219006360a0c0194f))


spec/all.js
var request = require('request-promise');
var API = require('../lib/mini-mock-api');

var basePath = '/api/v1';
var requestFromMockServer = function(path){
  return request({
    uri: 'http://localhost:8080' + basePath + path,
    json: true
  });
};

describe('File Router', function(){

  var server = new API({
    basePath: basePath,
    port: 8080,
    mockPath: 'example/mock-files',
    idAttribute: '_id',
  });

  beforeEach(function(){
    server.start();
  });

  afterEach(function(){
    server.stop();
  });

  it('should return the content of a JSON file if file exists', function(done){
    requestFromMockServer('/todos').then(function(response){
      expect(response.length).toBe(2);
      done();
    });
  });

  it('should return the object with the given id if file does not exist but route matches a file', function(done){
    requestFromMockServer('/todos/1').then(function(response){
      expect(response._id).toBe(1);
      done();
    });
  });

  it('should return a 404 if a file is not found', function(done){
    requestFromMockServer('/xyz').catch(function(error){
      expect(error.statusCode).toBe(404);
      done();
    });
  });

});


describe('Decorate Function', function(){

  var server = new API({
    basePath: basePath,
    port: 8080,
    mockPath: 'example/mock-files',
    idAttribute: '_id'
  });

  server.decorate = function(res, req){
    return {
      test: res
    };
  };

  beforeEach(function(){
    server.start();
  });

  afterEach(function(){
    server.stop();
  });

  it('should decorate the output if defined', function(done){
    requestFromMockServer('/todos/1').then(function(response){
      expect(response.test._id).toBe(1);
      done();
    });
  });
});

describe('CORS Option', function(){

  var server = new API({
    basePath: basePath,
    port: 8080,
    mockPath: 'example/mock-files',
    idAttribute: '_id',
    cors: true
  });

  beforeEach(function(){
    server.start();
  });

  afterEach(function(){
    server.stop();
  });

  it('should send CORS headers if cors option is enabled', function(done){
    request({
      uri: 'http://localhost:8080' + basePath + '/todos',
      json: true,
      transform: function(body, response){
        expect(response.headers['access-control-allow-origin']).toEqual('*');
        return body;
      }
    }).then(function(response){
      expect(response.length).toBe(2);
      done();
    });
  });
});


//   it('should sort the output if a sortParameter is defined', function(){
//
//   });
//
//   it('should sort the output if a custom sortParameter is defined', function(){
//
//   });
// });


example/example-cookie-based-login.js
var API = require('../lib/mini-mock-api');
var cookieParser = require('cookie-parser');

//session data
var sessionCookie = 'SESSION'
var sessionTimeout = 60*60*1000; // 1 hour
var username = 'test';
var password = 'test';

var server = new API({
  basePath: '/api/v1',
  port: 8080,
  mockPath: 'mock-files',
  idAttribute: '_id'
});
server.app.use(cookieParser());

//Login with /login?username=test&password=test
server.getFromRoot('/login', function (request, response) {
  if(request.query.username === username && request.query.password === password){
    response.cookie(sessionCookie, 'token', {expires: new Date(Date.now() + sessionTimeout)});
    response.status(200);
    response.json({status: 'successful'});
  } else {
    response.sendStatus(401);
  }
});

//Logout with /logout
server.getFromRoot('/logout', function (request, response) {
  response.clearCookie(sessionCookie);
  response.status(200);
  response.json({status: 'successful'});
});

//Ping /ping
server.getFromRoot('/ping', function (request, response) {
  if(request.cookies[sessionCookie]){
    response.cookie(sessionCookie, 'token', {expires: new Date(Date.now() + sessionTimeout)});
    response.json({user: username});
  } else {
    response.json({user: null});
  }
});

server.start();


lib/mini-mock-api.js
'use strict';

var express = require('express');
var fs = require('fs');
var path = require('path');
var _ = require('lodash-node');
var Q = require('q');
var fs_readFile = Q.denodeify(fs.readFile);
var customMethods = [
  'get', 'post', 'put', 'delete'
];

var MiniMockAPI = function(options){
  this.defaults = {
    mockPath: 'mocks',
    basePath: '/api',
    port: 8080,
    idAttribute: 'uuid',
    cors: false,
    sortParameter: 'sortOrder'
  };
  this.options = _.extend(this.defaults, options);
  this.app = express();
  if(this.options.cors === true){
    this.app.use(require('cors')());
  }
  if(this.options.delay){
    var options = {
      url: /./i,
      delay: 500
    };
    if(typeof this.options.delay === 'object'){
      _.extend(options, this.options.delay);
    }
    this.app.use(require('connect-slow')(options));
  }
  return this;
};


/*
* Handle all API requests with json files found in file system (./mocks).
* If a file is not found the server assumes that a single item was requested
* and tries to search a specific id in the parent collection
* e.g. for tasks/123 it will search in tasks.json for an item with idAttribute === 123
*
*/
MiniMockAPI.prototype.initMockRoute = function(filePath){
  this.app.get(filePath, function(request, response){
    var strippedRequestPath = request.path.substring(this.options.basePath.length),
    fileName = this.getFileName(strippedRequestPath);

    fs_readFile(fileName, 'utf8').then(

      function handleFileFound(data){
        var parsedData,
        responseData;
        try {
          parsedData = JSON.parse(data);
        } catch (e) {
          return console.error(e);
        }
        responseData = sort(parsedData, request.query[this.options.sortParameter]);
        console.log('Serving: ' + fileName)
        try {
          response.json(this.decorate(responseData, request));
        } catch (e) {
          console.log(e);
        }
      }.bind(this),

      function handleError(){
        var splitted = splitPathInIdAndPath(strippedRequestPath);
        if(splitted){
          fs_readFile(this.getFileName(splitted.path), 'utf8').then(

            function handleSingleItem(data){
              var parsedData,
              queryObj = {};
              queryObj[this.options.idAttribute] = splitted.id;

              try {
                parsedData = JSON.parse(data);
              } catch (e) {
                return console.error(e);
              }

              //search single item
              var foundItem = _.findWhere(parsedData, queryObj);
              //try again with number instead of string
              if(!foundItem){
                queryObj[this.options.idAttribute] = parseInt(splitted.id, 10);
                foundItem = _.findWhere(parsedData, queryObj);
              }
              if(foundItem){
                console.log('Serving: ' + path.join(this.getFileName(splitted.path), splitted.id));
                response.json(this.decorate(foundItem, request));
              } else {
                response.sendStatus(404);
              }
            }.bind(this),

            function(err) {
              console.log('No mock data found:', strippedRequestPath);
              response.sendStatus(404);
            }.bind(this)
          );
        }
      }.bind(this)
    );
  }.bind(this));
}

/*
* This function can be overridden. Allows to modify responses globally.
*/
MiniMockAPI.prototype.decorate = function(data) {
  return data;
};

/*
* Configure express mock route (has to be set up here because custom routes should be defined earlier)
* and start server
*/
MiniMockAPI.prototype.start = function() {
  this.initMockRoute(path.join(this.options.basePath, '*'));
  this.expressInstance = this.app.listen(this.options.port);
  console.log('Listening on:', this.options.port);
};

//Stop server
MiniMockAPI.prototype.stop = function() {
  if(this.expressInstance){
    this.expressInstance.close();
    console.log('Stopped listening on:', this.options.port);
  }
};

/*
* adds convenient functions for custom routes
* e.g. server.post('abc', function(req, res)) instead of server.app.post(server.options.basePath ...)
*/
customMethods.forEach(function(method){
  MiniMockAPI.prototype[method] = function(relativeRoute, fn) {
    this.app[method](path.join(this.options.basePath, relativeRoute), fn);
  };
  MiniMockAPI.prototype[method + 'FromRoot'] = function(absoluteRoute, fn) {
    this.app[method](absoluteRoute, fn);
  };
});


// get filename relative to mock path
MiniMockAPI.prototype.getFileName = function(filePath) {
  return path.join(process.cwd(), this.options.mockPath, filePath+'.json');
};

// split path in subpath and uuid
var splitPathInIdAndPath = function(filePath){
  var parts = filePath.split('/');
  if(parts.length > 0){
    var id = parts.pop();
    return {
      id: id,
      path: parts.join('/')
    }
  } else {
    return false;
  }
};

/*
* Sort data array by property
* leading + is ASC
* leading - is DESC
*/
var sort = function(data, sortOrder) {
  if(sortOrder){
    data = _.sortBy(data, sortOrder.substr(1));
    if(sortOrder[0] === '-'){
      data.reverse();
    }
  }
  return data;
};

module.exports = MiniMockAPI;
